libdnf/rpm/solv/goal_private.cpp

```diff
@@ -104,6 +129,141 @@ libdnf::rpm::solv::SolvMap list_results(
+    for (int i = 0; i < requires.size(); ++i) {
+        Id req_dep = requires[i];
+        Id p, pp;
+
+        FOR_PROVIDES(p, pp, req_dep)
+            if (p == b)
+                return true;
+    }
+
+    return false;
+}
+
+static void
+same_name_subqueue(Pool *pool, Queue * in, Queue * out)
+{
+    Id el = queue_pop(in);
+    Id name = pool_id2solvable(pool, el)->name;
+    queue_empty(out);
+    queue_push(out, el);
+    while (in->count &&
+           pool_id2solvable(pool, in->elements[in->count - 1])->name == name)
+        // reverses the order so packages are sorted by descending version
+        queue_push(out, queue_pop(in));
+}
+
+int sort_packages(const void *ap, const void *bp, void *s_cb)
+{
+    Id a = *(Id*)ap;
+    Id b = *(Id*)bp;
+    Pool *pool = ((struct InstallonliesSortCallback*) s_cb)->pool;
+    Id kernel = ((struct InstallonliesSortCallback*) s_cb)->running_kernel;
+    Solvable *sa = pool_id2solvable(pool, a);
+    Solvable *sb = pool_id2solvable(pool, b);
+
+    // if the names are different sort them differently, particular order does not matter as long as it's consistent.
+    int name_diff = sa->name - sb->name;
+    if (name_diff)
+        return name_diff;
+
+    // same name, if one is/depends on the running kernel put it last
+
+    // move available packages to end of the list
+    if (pool->installed != sa->repo)
+        return 1;
+
+    if (pool->installed != sb->repo)
+        return -1;
+
+    if (kernel >= 0) {
+        if (a == kernel || can_depend_on(pool, sa, kernel))
+            return 1;
+        if (b == kernel || can_depend_on(pool, sb, kernel))
+            return -1;
+        // if package has same evr as kernel try them to keep (kernel-devel packages)
+        Solvable * kernelSolvable = pool_id2solvable(pool, kernel);
+        if (sa->evr == kernelSolvable->evr) {
+            return 1;
+        }
+        if (sb->evr == kernelSolvable->evr) {
+            return -1;
+        }
+    }
+    return pool_evrcmp(pool, sa->evr, sb->evr, EVRCMP_COMPARE);
+}
+
+bool limit_installonly_packages(Solver * solv, libdnf::rpm::solv::IdQueue & job, const libdnf::rpm::solv::IdQueue & installonly, unsigned int installonly_limit,  Id running_kernel) {
+    if (installonly_limit == 0) {
+        return 0;
+    }
+
+    Pool * pool = solv->pool;
+    bool reresolve = false;
+
+    for (int i = 0; i < installonly.size(); ++i) {
+        Id p, pp;
+        libdnf::rpm::solv::IdQueue q;
+        libdnf::rpm::solv::IdQueue installing;
+        FOR_PROVIDES(p, pp, installonly[i]) {
+            // TODO(jmracek)  Replase the test by cached data from sack.p_impl->get_solvables()
+            if (!libdnf::utils::is_package(pool, p)) {
+                continue;
+            }
+            if (solver_get_decisionlevel(solv, p) > 0) {
+                q.push_back(p);
+            }
+        }
+        if (q.size() <= (int) installonly_limit) {
```

Comment:
Prefer a C++ style cast (in this case `static_cast<int>(installonly_limit)`) over a C style cast.
- C style cast is more dangerous. In fact C style cast `(type)variable` tries to do the following casts, in this order: `const_cast`, `static_cast`, `static_cast` followed by `const_cast`, `reinterpret_cast`, `reinterpret_cast` followed by `const_cast`
- It is easier to find C++ style casts in source code then C style casts.

It is on more places.